### ISSUE TITLE
Support installation in Virtualbox in EFI mode

### ISSUE THREAD

**Issue by tpgxyz** [calamares/calamares#839]
#### Submission type

  - [x] Bug report
  - [] Feature Request

#### Info regarding which version of Calamares is used, which Distribution

> 3.1.6

#### Provide information on how the disks are set up, in detail, with full logs of commands issued 

> …

#### What do you expect to have happen when Calamares installs?

> Calamares creates symlink to /EFI/BOOT/BOOTX64.EFI

#### Describe the issue you encountered

> Install ISO with Calamares on Virtualbox in EFI mode.  Boot installed system in EFI mode. Watch system not boot. Based on https://wiki.archlinux.org/index.php/VirtualBox#Installation_in_EFI_mode

#### Steps to reproduce the problem

> Install ISO with calamares in EFI mode.
> Boot installed system in EFI mode
> Observer Vbox whines from missing /EFI/BOOT/BOOTX64.EFI

#### Include the installation.log:

> …

**Comment by itchka** [comment]
To fill this out a bit. VirtualBox does not store efivars from one session to the next. If you set auto reboot in Calamares at the end the system will reboot but if you kill the install it will not. Currently the install EFI structure is this:-
efi/EFI
├── boot
│   └── grubx64.efi
└── openmandriva
    ├── AllocPages.efi
    ├── debughook.efi
    ├── drv0.efi
    ├── drv0_use.efi
    ├── exit.efi
    ├── FreePages.efi                                                                                                                                                                                                                                        
    ├── grub.cfg                                                                                                                                                                                                                                             
    ├── grub.efi                                                                                                                                                                                                                                             
    ├── grubx64.efi                                                                                                                                                                                                                                          
    ├── modelist.efi                                                                                                                                                                                                                                         
    ├── printenv.efi                                                                                                                                                                                                                                         
    ├── route80h.efi                                                                                                                                                                                                                                         
    ├── setjmp.efi                                                                                                                                                                                                                                           
    ├── t2.efi                                                                                                                                                                                                                                               
    ├── t3.efi                                                                                                                                                                                                                                               
    ├── t4.efi                                                                                                                                                                                                                                               
    ├── t5.efi                                                                                                                                                                                                                                               
    ├── t6.efi                                                                                                                                                                                                                                               
    ├── t7.efi                                                                                                                                                                                                                                               
    ├── t8.efi                                                                                                                                                                                                                                               
    ├── tcc.efi                                                                                                                                                                                                                                              
    └── t.efi    
But it needs to be this.
efi/EFI
├── boot
│   └── bootx64.efi
└── openmandriva
    ├── AllocPages.efi
    ├── debughook.efi
    ├── drv0.efi
    ├── drv0_use.efi
    ├── exit.efi
    ├── FreePages.efi
    ├── grub.cfg                   
    ├── grub.efi    
    ├── grubx64.efi
    ├── modelist.efi
    ├── printenv.efi                           
    ├── route80h.efi               
    ├── setjmp.efi     
    ├── t2.efi
    ├── t3.efi
    ├── t4.efi                                    
    ├── t5.efi                        
    ├── t6.efi            
    ├── t7.efi
    ├── t8.efi
    ├── tcc.efi                                      
    └── t.efi 
This allows VirtualBox to boot the install. An additional advantage that in the case of a machine with corrupted efivars (or no sane default) to boot without issue.

**Comment by itchka** [comment]
Tests have shown that the file can simple be renamed.

**Comment by oberon2007** [comment]
Confirming the issue on Manjaro.
Reboot after install works, new boot fails with EFI error.
Fixed by renaming to `grubx64.efi` to `bootx64.efi`

**Comment by tpgxyz** [comment]
Found the culprit. Output filenames got changed:
bootia32.efi -> grubia32.efi

https://github.com/calamares/calamares/commit/50cefe2ca80c3d6988e4dc5c0df1951a0244db69#diff-bfa2594912fd81abeb71d7a4b98c2905

**Comment by oberon2007** [comment]
@philmmanjaro , it's a bit strange. With your fix vbox boots the installed system successfully, however at the same time during boot it displays an error:

![efi_error](https://user-images.githubusercontent.com/11800714/31897932-15bb9724-b818-11e7-8b95-c1045ef7ac87.png)

**Comment by oberon2007** [comment]
Well, it's very weird @philmmanjaro, especially because apparently you say that you didn't have any issue in the first place - doesn't seem to make sense... How about you @tpgxyz @itchka ? Do you see my error?
Strange that it says DVD/CDROM while I have the media removed - even when I remove the virtual drive altogether the error is still displayed!

**Comment by itchka** [comment]
I see something flash by which could be that. It goes so fast though it's hard to tell

**Comment by adriaandegroot** [comment]
I'm reopening because of ongoing discussion on IRC, and we've not got a consensus whether the bug exists at all; and if so, what the right fix is.

**Comment by adriaandegroot** [comment]
Some history (assume 64-bitness here; I don't feel like writing out all the filenames twice):

1. There's a setting for *efi_file_source* ; you can see that this was `grubx64.efi` in the diff for 50cefe2ca80c3d6988e4dc5c0df1951a0244db69#diff-d10832f8a60f090d892d5e81093f83cbL234
2. There was, prior to that commit, a setting for *efi_file_target* ; it was `boot64.efi`
3. That commit simplified things, and introduced a bug: it copies *efi_file_source* into the EFI boot directory without renaming it. *efi_file_target* was dropped.
4. There are two different commits claiming to fix the problem:
   - PR #840 changes *efi_file_source* to be `bootx64.efi` ; that assumes that the source is (also) available under that name, and @philmmanjaro says that's wrong.
   - df84208abcbfd948da975602c9226f92cbb210a5 which got pushed to the tree re-introduces *efi_file_target*, and renames while copying.

Now there's a screenshot https://user-images.githubusercontent.com/7284530/31912695-a49177ce-b83c-11e7-825d-ce9e6d0e2212.png which claims to be from after step 4, there. Er .. @tpgxyz , it's not, since you can see the call to `shutil.copy2()` still has the variable `install_efi_boot_directory`, and the fact that it can't find the **source** file called `bootx64.efi` supports @philmmanjaro 's statement that that is the wrong source filename.

So, summing up, it looks like downstream is a mix-and-match of patches, which makes it really hard to see (like I said), if there is still a bug or not.

**Comment by adriaandegroot** [comment]
That said, any information merged into https://github.com/calamares/calamares/wiki/Tester's-Guide on setting up VBox for EFI boot is very welcome.

**Comment by philmmanjaro** [comment]
Well, installing grub you normally do it like this:

```
grub-install --target=x86_64-efi --efi-directory=/mnt/esp/ --boot-directory=/mnt/esp/ --bootloader-id=boot --modules=part_gpt
grub-install --target=i386-efi --efi-directory=/mnt/esp/ --boot-directory=/mnt/esp/ --bootloader-id=boot --modules=part_gpt
#grub-install /dev/sdd --boot-directory=/mnt/esp/ --bootloader-id=boot --modules=ntldr
mv /mnt/esp/EFI/boot/grubx64.efi /mnt/esp/EFI/boot/bootx64.efi
mv /mnt/esp/EFI/boot/grubia32.efi /mnt/esp/EFI/boot/bootia32.efi
grub-mkconfig -o /mnt/esp/grub/grub.cfg
rm -r /boot/grub/
```

This is how we do it on [Arch-based distros](http://www.taylorbyte.com/docs/wiki/archlinux/uefi-gpt-usb). If other distros do it another way we can discuss it.

**Comment by philmmanjaro** [comment]
@adriaandegroot: the issue with the screenshot posted by @itchka is simply cos he applied the patch by @tpgxyz which is wrong as there is no **bootx64.efi** as source. I already commented this and closed the PR once again.

**Comment by philmmanjaro** [comment]
On Manjaro side we have now the following:

```
[johnd@john-pc boot]$ tree .
.
├── efi
│   └── EFI
│       ├── boot
│       │   └── bootx64.efi
│       └── Manjaro
│           └── grubx64.efi
├── grub
│   ├── fonts
│   │   └── unicode.pf2
│   ├── grub.cfg
│   ├── grubenv
│   ├── locale
│   │   ├── ast.mo
│   │   ├── ca.mo
│   │   ├── da.mo
│   │   ├── de_CH.mo
│   │   ├── de@hebrew.mo
│   │   ├── de.mo
│   │   ├── en@arabic.mo
│   │   ├── en@hebrew.mo
│   │   ├── en@piglatin.mo
│   │   ├── en@quot.mo
│   │   ├── eo.mo
│   │   ├── es.mo
│   │   ├── fi.mo
│   │   ├── fr.mo
│   │   ├── gl.mo
│   │   ├── hr.mo
│   │   ├── hu.mo
│   │   ├── id.mo
│   │   ├── it.mo
│   │   ├── ja.mo
│   │   ├── ko.mo
│   │   ├── lt.mo
│   │   ├── nb.mo
│   │   ├── nl.mo
│   │   ├── pa.mo
│   │   ├── pl.mo
│   │   ├── pt_BR.mo
│   │   ├── ru.mo
│   │   ├── sl.mo
│   │   ├── sr.mo
│   │   ├── sv.mo
│   │   ├── tr.mo
│   │   ├── uk.mo
│   │   ├── vi.mo
│   │   ├── zh_CN.mo
│   │   └── zh_TW.mo
│   ├── themes
│   │   └── starfield
│   │       ├── blob_w.png
│   │       ├── boot_menu_c.png
│   │       ├── boot_menu_e.png
│   │       ├── boot_menu_ne.png
│   │       ├── boot_menu_n.png
│   │       ├── boot_menu_nw.png
│   │       ├── boot_menu_se.png
│   │       ├── boot_menu_s.png
│   │       ├── boot_menu_sw.png
│   │       ├── boot_menu_w.png
│   │       ├── COPYING.CC-BY-SA-3.0
│   │       ├── dejavu_10.pf2
│   │       ├── dejavu_12.pf2
│   │       ├── dejavu_14.pf2
│   │       ├── dejavu_16.pf2
│   │       ├── dejavu_bold_14.pf2
│   │       ├── README
│   │       ├── slider_c.png
│   │       ├── slider_n.png
│   │       ├── slider_s.png
│   │       ├── starfield.png
│   │       ├── terminal_box_c.png
│   │       ├── terminal_box_e.png
│   │       ├── terminal_box_ne.png
│   │       ├── terminal_box_n.png
│   │       ├── terminal_box_nw.png
│   │       ├── terminal_box_se.png
│   │       ├── terminal_box_s.png
│   │       ├── terminal_box_sw.png
│   │       ├── terminal_box_w.png
│   │       └── theme.txt
│   └── x86_64-efi
│       ├── acpi.mod
│       ├── adler32.mod
│       ├── affs.mod
│       ├── afs.mod
│       ├── ahci.mod
│       ├── all_video.mod
│       ├── aout.mod
│       ├── appleldr.mod
│       ├── archelp.mod
│       ├── ata.mod
│       ├── at_keyboard.mod
│       ├── backtrace.mod
│       ├── bfs.mod
│       ├── bitmap.mod
│       ├── bitmap_scale.mod
│       ├── blocklist.mod
│       ├── boot.mod
│       ├── boottime.mod
│       ├── bsd.mod
│       ├── bswap_test.mod
│       ├── btrfs.mod
│       ├── bufio.mod
│       ├── cacheinfo.mod
│       ├── cat.mod
│       ├── cbfs.mod
│       ├── cbls.mod
│       ├── cbmemc.mod
│       ├── cbtable.mod
│       ├── cbtime.mod
│       ├── chain.mod
│       ├── cmdline_cat_test.mod
│       ├── cmp.mod
│       ├── cmp_test.mod
│       ├── command.lst
│       ├── configfile.mod
│       ├── core.efi
│       ├── cpio_be.mod
│       ├── cpio.mod
│       ├── cpuid.mod
│       ├── crc64.mod
│       ├── cryptodisk.mod
│       ├── crypto.lst
│       ├── crypto.mod
│       ├── cs5536.mod
│       ├── ctz_test.mod
│       ├── datehook.mod
│       ├── date.mod
│       ├── datetime.mod
│       ├── diskfilter.mod
│       ├── disk.mod
│       ├── div.mod
│       ├── div_test.mod
│       ├── dm_nv.mod
│       ├── echo.mod
│       ├── efifwsetup.mod
│       ├── efi_gop.mod
│       ├── efinet.mod
│       ├── efi_uga.mod
│       ├── ehci.mod
│       ├── elf.mod
│       ├── eval.mod
│       ├── exfat.mod
│       ├── exfctest.mod
│       ├── ext2.mod
│       ├── extcmd.mod
│       ├── fat.mod
│       ├── file.mod
│       ├── fixvideo.mod
│       ├── font.mod
│       ├── fshelp.mod
│       ├── fs.lst
│       ├── functional_test.mod
│       ├── gcry_arcfour.mod
│       ├── gcry_blowfish.mod
│       ├── gcry_camellia.mod
│       ├── gcry_cast5.mod
│       ├── gcry_crc.mod
│       ├── gcry_des.mod
│       ├── gcry_dsa.mod
│       ├── gcry_idea.mod
│       ├── gcry_md4.mod
│       ├── gcry_md5.mod
│       ├── gcry_rfc2268.mod
│       ├── gcry_rijndael.mod
│       ├── gcry_rmd160.mod
│       ├── gcry_rsa.mod
│       ├── gcry_seed.mod
│       ├── gcry_serpent.mod
│       ├── gcry_sha1.mod
│       ├── gcry_sha256.mod
│       ├── gcry_sha512.mod
│       ├── gcry_tiger.mod
│       ├── gcry_twofish.mod
│       ├── gcry_whirlpool.mod
│       ├── geli.mod
│       ├── gettext.mod
│       ├── gfxmenu.mod
│       ├── gfxterm_background.mod
│       ├── gfxterm_menu.mod
│       ├── gfxterm.mod
│       ├── gptsync.mod
│       ├── grub.efi
│       ├── gzio.mod
│       ├── halt.mod
│       ├── hashsum.mod
│       ├── hdparm.mod
│       ├── hello.mod
│       ├── help.mod
│       ├── hexdump.mod
│       ├── hfs.mod
│       ├── hfspluscomp.mod
│       ├── hfsplus.mod
│       ├── http.mod
│       ├── iorw.mod
│       ├── iso9660.mod
│       ├── jfs.mod
│       ├── jpeg.mod
│       ├── keylayouts.mod
│       ├── keystatus.mod
│       ├── ldm.mod
│       ├── legacycfg.mod
│       ├── legacy_password_test.mod
│       ├── linux16.mod
│       ├── linux.mod
│       ├── loadbios.mod
│       ├── loadenv.mod
│       ├── loopback.mod
│       ├── lsacpi.mod
│       ├── lsefimmap.mod
│       ├── lsefi.mod
│       ├── lsefisystab.mod
│       ├── lsmmap.mod
│       ├── ls.mod
│       ├── lspci.mod
│       ├── lssal.mod
│       ├── luks.mod
│       ├── lvm.mod
│       ├── lzopio.mod
│       ├── macbless.mod
│       ├── macho.mod
│       ├── mdraid09_be.mod
│       ├── mdraid09.mod
│       ├── mdraid1x.mod
│       ├── memdisk.mod
│       ├── memrw.mod
│       ├── minicmd.mod
│       ├── minix2_be.mod
│       ├── minix2.mod
│       ├── minix3_be.mod
│       ├── minix3.mod
│       ├── minix_be.mod
│       ├── minix.mod
│       ├── mmap.mod
│       ├── moddep.lst
│       ├── modinfo.sh
│       ├── morse.mod
│       ├── mpi.mod
│       ├── msdospart.mod
│       ├── mul_test.mod
│       ├── multiboot2.mod
│       ├── multiboot.mod
│       ├── nativedisk.mod
│       ├── net.mod
│       ├── newc.mod
│       ├── nilfs2.mod
│       ├── normal.mod
│       ├── ntfscomp.mod
│       ├── ntfs.mod
│       ├── odc.mod
│       ├── offsetio.mod
│       ├── ohci.mod
│       ├── part_acorn.mod
│       ├── part_amiga.mod
│       ├── part_apple.mod
│       ├── part_bsd.mod
│       ├── part_dfly.mod
│       ├── part_dvh.mod
│       ├── part_gpt.mod
│       ├── partmap.lst
│       ├── part_msdos.mod
│       ├── part_plan.mod
│       ├── part_sun.mod
│       ├── part_sunpc.mod
│       ├── parttool.lst
│       ├── parttool.mod
│       ├── password.mod
│       ├── password_pbkdf2.mod
│       ├── pata.mod
│       ├── pbkdf2.mod
│       ├── pbkdf2_test.mod
│       ├── pcidump.mod
│       ├── play.mod
│       ├── png.mod
│       ├── priority_queue.mod
│       ├── probe.mod
│       ├── procfs.mod
│       ├── progress.mod
│       ├── raid5rec.mod
│       ├── raid6rec.mod
│       ├── random.mod
│       ├── read.mod
│       ├── reboot.mod
│       ├── regexp.mod
│       ├── reiserfs.mod
│       ├── relocator.mod
│       ├── romfs.mod
│       ├── scsi.mod
│       ├── search_fs_file.mod
│       ├── search_fs_uuid.mod
│       ├── search_label.mod
│       ├── search.mod
│       ├── serial.mod
│       ├── setjmp.mod
│       ├── setjmp_test.mod
│       ├── setpci.mod
│       ├── sfs.mod
│       ├── shift_test.mod
│       ├── signature_test.mod
│       ├── sleep.mod
│       ├── sleep_test.mod
│       ├── spkmodem.mod
│       ├── squash4.mod
│       ├── strtoull_test.mod
│       ├── syslinuxcfg.mod
│       ├── tar.mod
│       ├── terminal.lst
│       ├── terminal.mod
│       ├── terminfo.mod
│       ├── test_blockarg.mod
│       ├── testload.mod
│       ├── test.mod
│       ├── testspeed.mod
│       ├── tftp.mod
│       ├── tga.mod
│       ├── time.mod
│       ├── trig.mod
│       ├── tr.mod
│       ├── true.mod
│       ├── udf.mod
│       ├── ufs1_be.mod
│       ├── ufs1.mod
│       ├── ufs2.mod
│       ├── uhci.mod
│       ├── usb_keyboard.mod
│       ├── usb.mod
│       ├── usbms.mod
│       ├── usbserial_common.mod
│       ├── usbserial_ftdi.mod
│       ├── usbserial_pl2303.mod
│       ├── usbserial_usbdebug.mod
│       ├── usbtest.mod
│       ├── verify.mod
│       ├── video_bochs.mod
│       ├── video_cirrus.mod
│       ├── video_colors.mod
│       ├── video_fb.mod
│       ├── videoinfo.mod
│       ├── video.lst
│       ├── video.mod
│       ├── videotest_checksum.mod
│       ├── videotest.mod
│       ├── xfs.mod
│       ├── xnu.mod
│       ├── xnu_uuid.mod
│       ├── xnu_uuid_test.mod
│       ├── xzio.mod
│       ├── zfscrypt.mod
│       ├── zfsinfo.mod
│       └── zfs.mod
├── initramfs-4.14-x86_64-fallback.img
├── initramfs-4.14-x86_64.img
├── intel-ucode.img
├── linux414-x86_64.kver
├── memtest86+
│   └── memtest.bin
└── vmlinuz-4.14-x86_64
```

**Comment by philmmanjaro** [comment]
@oberon2007: since I had an CD mounted I didn't got the error message displayed. Now I do. However it still boots. Having it with **grubx64.efi** I can reproduce it not booting. So with my given fix it works at least for **Manjaro**, minus the weird `Boot failed. EFI DVD/CDROM` message.

**Comment by oberon2007** [comment]
yes. that's exactly what I get.

**Comment by philmmanjaro** [comment]
@oberon2007: I assume VB remembers that we had booted from DVD and then simply tries that first. Might be possible to change that within the EFI bios of VB not to try DVD first. However I see no issue with it at all.

**Comment by oberon2007** [comment]
I have unchecked 'Optical' in vbox' system settings, and even when removing the virtual drive from the setup, the error keeps showing for me.
Anyway I agree that this is not serious and most likely not a calamares issue.

**Comment by philmmanjaro** [comment]
Well, you can actually change the boot order or remove the optical option within the BIOS of VB.

![virtualbox_manjaro_24_10_2017_07_28_25](https://user-images.githubusercontent.com/5050022/31940730-256611d4-b88d-11e7-8c36-b006b478b588.png)

**Comment by itchka** [comment]
Ok Guys my apologies I seem to ghave got my patches mixed up. Time for some early nights I think.
Everything works great now
